src/ride/ride.c

```diff
@@ -6171,9 +6170,47 @@ void ride_update_max_vehicles(int rideIndex)
 	}
 }
 
-static void sub_6DE52C(rct_ride *ride)
+/*
+ * rtc2: 0x006DE52C
+ */
+static void ride_reset_vehicle_colours(rct_ride *ride)
 {
-	RCT2_CALLPROC_X(0x006DE52C, 0, 0, 0, 0, (int)ride, 0, 0);
+	uint8* colourCount = GET_RIDE_ENTRY(ride->subtype)->default_colours_ptr;
+	rct_vehicle_colour_extended* colour = colourCount + 1;
+
+	//Different colour per train
+	if (*colourCount == 255) {
+		ride->colour_scheme_type = RIDE_COLOUR_SCHEME_DIFFERENT_PER_TRAIN;
+		for (int index = 0; index < 32; index++) {
+			ride->vehicle_colours[index] = colour[index].vehicle_colour;
+			ride->vehicle_colours_extended[index] = colour[index].restraint_colour;
+		}
+		return;
+	}
+
+	//Same colour per train, different from all existing rides of this type if possible
+	ride->colour_scheme_type = RIDE_COLOUR_SCHEME_ALL_SAME;
+	int remaining = 0xC8;		//Not sure why 200
+	rct_ride* ride_list = RCT2_ADDRESS(RCT2_ADDRESS_RIDE_LIST, rct_ride);
+	uint16 colourIndex;
+	rct_ride* rideTemp;
+
+getRandomColorScheme:
```

Comment:
I'm not the biggest fan of making this a goto. You could make it a for loop that tries 200 times and if it is succesful returns after setting the colour and if unsuccessful return with colourIndex 0.